File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
import collections
sys.setrecursionlimit(10 ** 6)

N, M = map(int, sys.stdin.readline().split(" "))

dx = [0, 1, 0, -1] # 여기부분 아래, 혹은 오른쪽 방향으로 우선순위하도록 맞추기
dy = [1, 0, -1, 0]

arr = [[0]]

for i in range(N):
    num1 = [0]
    num2 = list(map(int, sys.stdin.readline().rstrip()))
    arr.append(num1 + num2)
# 좌표가 1, 1부터 시작하기에 일부로 N, M이 0 0 인 자리를 생각하지 않도록 만드는 과정 진행

count = 0 # 최소의 칸수 

def BFS(y, x):
    q = collections.deque([(y, x)])
    while q:
        startY , startX = q.popleft()
        for s in range(4):
            newY = startY + dy[s]
            newX = startX + dx[s]
            if(newX > 0 and newX <= M  and newY > 0 and newY <= N):
                if(arr[newY][newX] == 1):
                    arr[newY][newX] = arr[startY][startX] + 1
                    q.append((newY,newX))

BFS(1, 1) # 좌표 1, 1부터 시작
print(arr[N][M])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.io.*;


public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int[] dx = {0, 0, 1, -1};
    private static int[] dy = {1, -1, 0, 0};
    private static int N, M, count;
    private static int[][] arr;
    private static boolean[][] visited;

    private static void BFS(int y, int x) {
        Deque<int[]> q = new LinkedList<>();
        visited[y][x] = true;
        q.add(new int[] {y, x});
        while(!q.isEmpty()) {
            int[] dots = q.poll();
            for(int i = 0 ; i < 4 ; i++) {
                int newX = dots[1] + dx[i];
                int newY = dots[0] + dy[i];
                if(newX >= 0 && newX < M && newY >= 0 && newY < N) {
                    if(arr[newY][newX] != 0 && visited[newY][newX] == false) {
                        visited[newY][newX] = true;
                        arr[newY][newX] = arr[dots[0]][dots[1]] + 1;
                        q.add(new int[] {newY, newX}); 
                    }
                }
            }
        }
    }



    public static void main(String[] args) throws IOException {
        
        st = new StringTokenizer(br.readLine(), " ");

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N][M];
        visited = new boolean[N][M];

        for(int i = 0 ; i < N ; i++){
            String nums = br.readLine();
            for(int j = 0 ; j < nums.length() ; j++) {
                int num = nums.charAt(j) - '0';
                arr[i][j] = num;
            }
        }

        BFS(0, 0);
        
        bw.write(Integer.toString(arr[N - 1][M - 1]));

        bw.flush();

        br.close();
        bw.close();
    }
}